test/slam_manifold_preintegration_update_test.cc

```diff
@@ -0,0 +1,303 @@
+                 Ref<Vector3d> new_Dp) {
+  const auto corrected_accel = accel - accel_bias;
+  new_DR = DR * Rot3d::FromTangent((gyro - gyro_bias) * dt, epsilon);
+  new_Dv = Dv + DR * corrected_accel * dt;
+  new_Dp = Dp + Dv * dt + DR * corrected_accel * (0.5 * dt * dt);
+}
+
+TEST_CASE("Test ImuManifoldPreintegrationUpdate has correct DR, Dv, & Dp", "[slam]") {
+  std::mt19937 gen(1804);
+  std::normal_distribution<double> dist{0.0, 10.0};
+  const auto randomM3 = Eigen::Matrix3d::NullaryExpr([&]() { return dist(gen); });
+  const auto randomV3 = Eigen::Vector3d::NullaryExpr([&]() { return dist(gen); });
+  const auto randomM9 = Eigen::Matrix<double, 9, 9>::NullaryExpr([&]() { return dist(gen); });
+
+  for (int i_ = 0; i_ < 10; i_++) {
+    const sym::Rot3d DR = sym::Rot3d::Random(gen);
+    const Eigen::Vector3d Dv = randomV3;
+    const Eigen::Vector3d Dp = randomV3;
+    const Eigen::Matrix<double, 9, 9> covariance = randomM9;
+    const Eigen::Matrix3d DR_D_gyro_bias = randomM3;
+    const Eigen::Matrix3d Dv_D_gyro_bias = randomM3;
+    const Eigen::Matrix3d Dv_D_accel_bias = randomM3;
+    const Eigen::Matrix3d Dp_D_gyro_bias = randomM3;
+    const Eigen::Matrix3d Dp_D_accel_bias = randomM3;
+
+    const Eigen::Vector3d gyro = randomV3;
+    const Eigen::Vector3d accel = randomV3;
+
+    const double dt = 1.24;
+
+    const Eigen::Vector3d gyro_bias = randomV3;
+    const Eigen::Vector3d accel_bias = randomV3;
+    const Eigen::Matrix3d gyro_cov = randomM3;
+    const Eigen::Matrix3d accel_cov = randomM3;
+
+    sym::Rot3d new_DR;
+    Eigen::Vector3d new_Dv;
+    Eigen::Vector3d new_Dp;
+    Eigen::Matrix<double, 9, 9> new_covarinace;
+    Eigen::Matrix3d new_DR_D_gyro_bias;
+    Eigen::Matrix3d new_Dp_D_gyro_bias;
+    Eigen::Matrix3d new_Dp_D_accel_bias;
+    Eigen::Matrix3d new_Dv_D_gyro_bias;
+    Eigen::Matrix3d new_Dv_D_accel_bias;
+
+    sym::ImuManifoldPreintegrationUpdate<double>(
+        DR, Dv, Dp, covariance, DR_D_gyro_bias, Dv_D_gyro_bias, Dv_D_accel_bias, Dp_D_gyro_bias,
+        Dp_D_accel_bias, gyro_bias, accel_bias, gyro_cov, accel_cov, gyro, accel, dt,
+        sym::kDefaultEpsilond, &new_DR, &new_Dv, &new_Dp, nullptr, nullptr, nullptr, nullptr,
+        nullptr, nullptr);
+
+    sym::Rot3d expected_DR;
+    Eigen::Vector3d expected_Dv;
+    Eigen::Vector3d expected_Dp;
+
+    UpdateState(DR, Dv, Dp, gyro, accel, gyro_bias, accel_bias, dt, sym::kDefaultEpsilond,
+                expected_DR, expected_Dv, expected_Dp);
+
+    CHECK(sym::IsClose(new_DR, expected_DR, 1e-14));
+    CHECK(sym::IsClose(new_Dv, expected_Dv, 1e-14));
+    CHECK(sym::IsClose(new_Dp, expected_Dp, 1e-14));
+  }
+}
+
+/**
+ * Helper class to generate samples from a multi-variate normal distribution with
+ * the same covariance as that passed into the constructor.
+ */
+class MultiVarNormalDist {
+ private:
+  const Eigen::MatrixXd L;
+  const long int vars;
```

Comment:
Well, the compiler gave a warning of a narrowing conversion if I used int, but I think the code is more straightforward if I just use `L.cols()` instead, so that's what I've done.